src/lib/statistics/cardinality_estimator.cpp

```diff
@@ -0,0 +1,911 @@
+            const auto total_distinct_count =
+                std::max(scan_statistics_object->total_distinct_count(), HistogramCountType{1.0f});
+            selectivity = total_distinct_count > 0 ? (total_distinct_count - 1.0f) / total_distinct_count : 0.0f;
+          } break;
+
+          case PredicateCondition::LessThan:
+          case PredicateCondition::LessThanEquals:
+          case PredicateCondition::GreaterThan:
+          case PredicateCondition::GreaterThanEquals:
+          case PredicateCondition::BetweenInclusive:
+          case PredicateCondition::BetweenExclusive:
+          case PredicateCondition::BetweenLowerExclusive:
+          case PredicateCondition::BetweenUpperExclusive:
+          case PredicateCondition::In:
+          case PredicateCondition::NotIn:
+          case PredicateCondition::Like:
+          case PredicateCondition::NotLike:
+            // Lacking better options, assume a "magic" selectivity for >, >=, <, <=, ... Any number would be equally
+            // right and wrong here. In some examples, this seemed like a good guess ¯\_(ツ)_/¯
+            selectivity = 0.5f;
+            break;
+
+          case PredicateCondition::IsNull:
+          case PredicateCondition::IsNotNull:
+            Fail("IS (NOT) NULL predicates should not have a 'value' parameter.");
+        }
+
+      } else {
+        /**
+         * Estimate ColumnVsValue
+         */
+
+        Assert(predicate.value.type() == typeid(AllTypeVariant), "Expected AllTypeVariant");
+
+        // TODO(anybody) For (NOT) LIKE predicates that start with a wildcard, Histograms won't yield reasonable
+        //               results. Assume a magic selectivity for now
+        if (predicate.predicate_condition == PredicateCondition::Like) {
+          // Lacking better options, assume a "magic" selectivity for LIKE. Any number would be equally
+          // right and wrong here. In some examples, this seemed like a good guess ¯\_(ツ)_/¯
+          selectivity = 0.1f;
+          return;
+        }
+        if (predicate.predicate_condition == PredicateCondition::NotLike) {
+          // Lacking better options, assume a "magic" selectivity for NOT LIKE. Any number would be equally
+          // right and wrong here. In some examples, this seemed like a good guess ¯\_(ツ)_/¯
+          selectivity = 0.9f;
+          return;
+        }
+
+        auto value2_variant = std::optional<AllTypeVariant>{};
+        if (predicate.value2) {
+          if (predicate.value2->type() != typeid(AllTypeVariant)) {
+            selectivity = 1.0f;
+            return;
+          }
+
+          value2_variant = boost::get<AllTypeVariant>(*predicate.value2);
+        }
+
+        const auto sliced_statistics_object = scan_statistics_object->sliced(
+            predicate.predicate_condition, boost::get<AllTypeVariant>(predicate.value), value2_variant);
+
+        if (!sliced_statistics_object) {
+          selectivity = 0.0f;
```

Comment:
Why is this 0 while row 585 returns 1? Aren't these two cases where we don't know what to do? I'd still like for "dunno" to be coded more specifically.

Reply:
Okay, let's go crazy. How about
```c++
// Constants used in places where a better estimation would be implementable, either with
// unavailable statistics objects or new algorithms, but wasn't warranted yet
constexpr auto PLACEHOLDER_SELECTIVITY_ONE = 1.0f
constexpr auto PLACEHOLDER_SELECTIVITY_ZERO = 1.0f
constexpr auto PLACEHOLDER_SELECTIVITY_LOW = 0.1f
constexpr auto PLACEHOLDER_SELECTIVITY_MEDIUM = 0.5f
constexpr auto PLACEHOLDER_SELECTIVITY_HIGH = 0.5f
```

Reply:
I’d use “all” and “none” instead of one and zero. Also it seems like you adjusted only half of those numbers (zero, high).

Reply:
Using these in all places where a magic selectivity constant was used before.

Reply:
Except for line 589 (`selectivity = 0.0f;`)

Reply:
Actually, there are a couple more. Better to grep for `selectivity = (0|1)`

Reply:
For those four instances of `selectivity = 0.0f`, 0.0f is actually the intended estimation and not a placeholder for "we do not really handle this predicate"

589, e.g. is doc'ed like this:

```c++
          // A predicate `<column> <condition> NULL` always has a selectivity of 0
          selectivity = 0.0f;
```

I'd not use the PLACEHOLDER_ constants for such cases, no?

Reply:
Ok.